Research/Research/RSDTaskViewModel.swift

```diff
@@ -0,0 +1,947 @@
+    open func goForward() {
+        guard let viewModel = self.currentTaskPath as? RSDTaskViewModel, viewModel == self
+            else {
+                self.currentTaskPath.moveForwardToNextStep()
+                return
+        }
+        _goForward()
+    }
+    
+    private func _goForward() {
+        
+        guard self.task != nil else {
+            _fetchTaskFromCurrentInfo()
+            return
+        }
+        
+        // Add or update the result end date if there was previous step
+        if let previousStep = self.currentChild {
+            if let result = self.taskResult.stepHistory.last, result.identifier == previousStep.identifier {
+                var finalResult = result
+                finalResult.endDate = Date()
+                self.taskResult.appendStepHistory(with: finalResult)
+            }
+            else {
+                let result = previousStep.pathResult()
+                self.taskResult.appendStepHistory(with: result)
+            }
+        }
+        
+        // move to the next step
+        _moveToNextStep()
+    }
+    
+    /// Go back to the previous step.
+    ///
+    /// - note: This method will throw an assertion if there isn't a previous step.
+    open func goBack() {
+        guard let _ = self.currentChild else {
+            assertionFailure("Cannot go backward with a nil current step. path = \(self)")
+            return
+        }
+        
+        // If calling `goBack()` from the `taskViewController.taskViewModel`, that node is always the root
+        // node, and may not be the node that is currently being navigated.
+        guard let viewModel = self.currentTaskPath as? RSDTaskViewModel, viewModel == self
+            else {
+                self.currentTaskPath.moveBackToPreviousStep()
+                return
+        }
+        
+        // move to the previous step
+        _moveToPreviousStep()
+    }
+    
+    /// Call through to the task controller to handle the task finished with a reason of `.cancelled`.
+    open func cancel(shouldSave: Bool = false) {
+        let reason: RSDTaskFinishReason = shouldSave ? .saved : .discarded
+        self.taskController?.handleTaskDidFinish(with: reason, error: nil)
+    }
+    
+    /// Start the task if it is not currently loaded with a task or first step.
+    open func startTaskIfNeeded() {
+        guard !self.isLoading else { return }
+        if self.task == nil {
+            _fetchTaskFromCurrentInfo()
+        }
+        else if self.currentStep == nil {
+            _moveToNextStep()
+        }
+    }
+    
+    /// For the given step, what is the next path component and step controller (if applicable) for this step.
+    /// The base class implementation will return an `RSDTaskStepNode` for either a subtask step or a section
+    /// step. For all other steps, it will request a step controller for the step from the task controller and
+    /// return both the step controller with the loaded step view model.
```

Comment:
both the step controller... and _what_ ...with the loaded step view model? (or drop the 'both'? or change the 'with' to 'and'? or...?)